Biology.java

```diff
@@ -0,0 +1,397 @@
+import java.util.Map;
+import java.util.Scanner;
+
+/**
+ * A class that allows antigen to model a virus's genetic sequence,
+ * and how changes in sequence give rise to changes in both:
+ * i) antigenic phenotype
+ * ii) viral fitness in the absence of immune selection (e.g., quantified by DMS data)
+ *
+ * @author Thien Tran
+ */
+public class Biology {
+    public enum AlphabetType {
+        NUCLEOTIDES("ACGT"),
+        AMINO_ACIDS("ACDEFGHIKLMNPQRSTWYV");
+
+        private final String validCharacters;
+
+        AlphabetType(String validCharacters) {
+            this. validCharacters = validCharacters;
+        }
+
+        public String getValidCharacters() {
+            return validCharacters;
+        }
+    }
+
+    public enum DMSData {
+        DMS_DATA();
+        public double[][] aminoAcidPreference = null;
+        private final double[] equalPreferences = new double[]{0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05, 0.05};;
+
+        DMSData() {
+            try {
+                // Only initialize aminoAcidPreference if DMS data is provided.
+                if (Parameters.DMSFile != null) {
+                    int numberOfAminoAcidSites = Parameters.startingSequence.length() / 3;
+                    int numberOfAminoAcids = Biology.AlphabetType.AMINO_ACIDS.getValidCharacters().length();
+                    aminoAcidPreference = new double[numberOfAminoAcidSites][numberOfAminoAcids];
+
+                    Scanner dms = new Scanner(new File(Parameters.DMSFile));
+                    dms.nextLine(); // read header
+
+                    for (int i = 0; i < numberOfAminoAcidSites; i++) {
+                        Scanner currentSite = new Scanner(dms.nextLine());
+                        currentSite.useDelimiter(",");
+
+                        currentSite.next(); // ignore site number
+
+                        for (int j = 0; j < numberOfAminoAcids; j++) {
+                            double probability = currentSite.nextDouble();
+                            aminoAcidPreference[i][j] = probability;
+                        }
+                    }
+                }
+            } catch (Exception e) {
+                e.printStackTrace();
+            }
+        }
+
+        public double[] getAminoAcidPreference(int site) {
+            // If DMS data is not provided, then return double[] of equal amino acid preferences.
+            if (aminoAcidPreference == null) {
+                return equalPreferences;
+            }
+            return aminoAcidPreference[site];
+        }
+    }
+
+    /**
+     * Transition/transversion ratio used to determine what a given wild type nucleotide should be mutated to.
+     */
+    public enum MutationType {
+        MUTATION();
+        public final Map<Character, double[]> transitionTranversionProbability;
+
+        /**
+         * Constructor that creates a probability distribution for each nucleotide,
```

Comment:
Hugh explained that this is the array of "boundaries", which is not a probability matrix , for which I would have expected something more like https://en.wikipedia.org/wiki/Stochastic_matrix#:~:text=In%20mathematics%2C%20a%20stochastic%20matrix,substitution%20matrix%2C%20or%20Markov%20matrix.